# Practicing analysis pipelines for circRNA-seq data
## Introduction
### Background
circRNA is a non-coding RNA(ncRNA) that forms covalently closed loop and often acts as a gene regulator in mammals. It is generated by backsplicing of exons and lariat introns (Fig. 1 in [Ref1](https://www.frontiersin.org/articles/10.3389/fmolb.2017.00038/full)). circRNAs are evolutionary conserved and known to be relatively stable in cytoplasms. Its molecular function is acting as miRNA sponge or binding to RNA-associated proteins. The identification of circRNA is possible from typical ncRNAseq data. The target sequencing requires a specific experimental procedure and computational method to enrich or identify circRNAs from typical cytoplasmic ncRNAs.

![image](https://www.frontiersin.org/files/Articles/265463/fmolb-04-00038-HTML/image_m/fmolb-04-00038-g001.jpg)

Fig. 1 in [Ref1](https://www.frontiersin.org/articles/10.3389/fmolb.2017.00038/full) circRNA splicing. circRNAs are created by non-canonical splicing process known as “backsplicing.” A downstream splice donor is joined to an upstream splice acceptor. circRNAs can be exonic, intronic or a combination of both. Colored bars = exons. Black lines = introns.

In this class project, I will download the public ncRNAseq data from [SRA NCBI]() and practice the processing of data using the workflow of cirRNA-seq analysis is in the following figure of [Ref2](). The raw data is available from the [SRA]() database under the [Bioproject ID: PRJNA1019680](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA1019680). The analytic procedure in the original publication is followed.

```
The paired-end raw reads were trimmed, and the quality was controlled by Trimmomatic (v0.39)18
and FastQC (v0.11.9) with default parameters.
The Hisat2 v2.0.5 tool was used to build the index of the reference genome
and align paired-end clean reads to the reference genome.
Hisat2 was chosen as the mapping tool due to its capability to generate a splice junction database
based on the gene model annotation file, which provided improved mapping results.
We used find_circ19 and CIRI220 to detect and identify circRNAs.
First, BWA21 was used to compare the clean data with the reference genome.
Then, CIRI scans SAM files and detects junction reads with paired chiastic clipping (PCC) signals,
paired end mapping (PEM) and GT-AG splicing signals.
Find_circ obtains junction reads via Bowtie222.
The circos figure was constructed utilizing the Circos software.
The DESeq2 R package (1.20.0) was utilized to perform differential expression analysis
for two conditions/groups (with two biological replicates per condition).
```

While the paper are used different mapping programs (such as HISAT2, BWA, and Bowtie2) and two pipelines (e.g. find_circ and CIRI), I will use the most current version of [`circRNAprofiler`](https://bioconductor.org/packages/release/bioc/vignettes/circRNAprofiler/inst/doc/circRNAprofiler.html) from the [Bioconductor]() package. The `circRNAprofiler` package relies on the pre-computed circRNA identification using popular circRNA annotation pipelines. The overview of the `circRNAprofiler` procedure is shown in the [`circRNAprofiler`](https://bioconductor.org/packages/release/bioc/vignettes/circRNAprofiler/inst/doc/circRNAprofiler.html) tutorial.

To do this `circRNAprofiler` package or any discovery pipelines, the first step is to map QCed reads to the reference genome (here, human genome). The mapped reads are recorded in files (e.g. `BAM` [format](https://hhj6212.github.io/biology/tech/2020/08/26/Bioinformatics-fileformats.html)). Once `BAM` files were made, I can move to the next step (annotation of circRNAs). The typical read mapping procedure can be found in elsewhere.

## Methods
There are two ways of running pipeline. The first option is running required programs manually line by line. It requires skills for handling coding environment such as `conda`.

### (1) commandline pipeline
1. Download sample sequence (reads) data from the NCBI SRA database. See, how to [download](https://github.com/ncbi/sra-tools/wiki) using a SRA-tools kit.
2. For the practice, we will download a sample data. A sample data includes 6 samples (SRR id; 3 normal + 3 hypoxia)
[SRA data](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP462463&o=acc_s%3Aa&s=SRR26133767,SRR26133768,SRR26133769,SRR26133770,SRR26133771,SRR26133772)
<!-- 
- GSM7794632: HUVECs treatment with intersectant hypoxia [IH1]; Homo sapiens; ncRNA-Seq
https://www.ncbi.nlm.nih.gov/sra/SRX21847130[accn]
- GSM7794633: HUVECs treatment with intersectant hypoxia [IH2]; Homo sapiens; ncRNA-Seq
https://www.ncbi.nlm.nih.gov/sra/SRX21847131[accn]
- GSM7794634: HUVECs treatment with intersectant hypoxia [IH3]; Homo sapiens; ncRNA-Seq
https://www.ncbi.nlm.nih.gov/sra/SRX21847132[accn]
- GSM7794629: HUVECs treatment with control [CK1]; Homo sapiens; ncRNA-Seq
https://www.ncbi.nlm.nih.gov/sra/SRX21847127[accn]
- GSM7794630: HUVECs treatment with control [CK2]; Homo sapiens; ncRNA-Seq
https://www.ncbi.nlm.nih.gov/sra/SRX21847128[accn]
- GSM7794631: HUVECs treatment with control [CK3]; Homo sapiens; ncRNA-Seq
https://www.ncbi.nlm.nih.gov/sra/SRX21847129[accn]
-->

4. An easy way of installing `SRA-tools` is using `mini-conda`.
Install mini-conda as described in [the class title page](https://github.com/igchoi/IBT618-SystemsBiotechnology/tree/main). Then, install `sra-tools` under `circ` environment of `conda`. The creation and activation of `circ` environement.

```
conda create -n circ
conda activate circ
```
Install `sra-tools`, `bowties2` and all required programs.
```
conda install sra-tools
conda install bowtie2
```
`Bowtie2` program requires pre-indexed reference genome. Please, check the instuction for the mapping example - from indexing reference genome to mapping ([phage genome example](https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example))
Once read mapping is done, you can use a circRNA annotation program such as `finc_circ` that can be also installed using `conda`. 
```
conda install bioconda::find_circ
```
Please, check how to find circRNA using find_circ [here](https://github.com/marvin-jens/find_circ).

### (2) using `Galaxy` server
In Galaxy server, we can directly download reads data and mapping the reads into the reference genome. Please, check this GTN(galaxy tutorial network)::[transcriptomics](https://training.galaxyproject.org/training-material/topics/transcriptomics/) and [public workflow](https://usegalaxy.org/u/igchoi/w/unnamed-workflow-1) built during the class hours. The galaxy toolshed provides some [tools](https://toolshed.tools/#repos?q=circular%20rna) for circular RNA analysis. For example, this [tool::CIRCexplorer2]((https://toolshed.g2.bx.psu.edu/repository?repository_id=b30488c37fdc9083&changeset_revision=465d6578b4d0)) seems working for circular RNA analysis.


#### Questions to be solved by the class project
* Is it for discovery of new circRNAs? no
* Is it for identification of differentially expressed circRNAs? yes
 
#### Links
1. [갤럭시 서버::RNA-seq raw data 부터 분석하는 방법 - 서버 없이!](https://hhj6212.github.io/pipeline/informatics/2021/03/07/RNAseq-pipeline.html)
2. [RU_RNAseq tutorial using Bioconductor packages](https://rockefelleruniversity.github.io/RU_RNAseq/)
3. [CircRNAprofiler](https://bioconductor.org/packages/release/bioc/html/circRNAprofiler.html)
4. [published paper](https://www.nature.com/articles/s41598-024-51471-3#Sec2)
