# Class project
## Title: Developing analysis pipeline for understanding microbial diversity of Chinese Daqu samples
-----
### Background
* What is __Daqu__?
 ![88787374c6a11df817564da02f9a13d9](https://github.com/igchoi/IBT618-SystemsBiotechnology/assets/165759135/15a73096-b5dd-49f4-90e9-fb4de5dd1289)
* What are the applications of microbial diversity sequencing?
![7a2e2ddf211e9640cadf9c6e98f6fc67](https://github.com/igchoi/IBT618-SystemsBiotechnology/assets/165759135/c8303427-5d93-4c4e-b65e-46c1ea9cbee6)
* How to detect microbial diversity in the Daqu sample?
 ![fe9d909e3b1e84b0888b67d6e2271f2f](https://github.com/igchoi/IBT618-SystemsBiotechnology/assets/165759135/aff5810f-86ae-48c2-b464-fd8e82849d3c)
----
### Questions to be solved by the class project
* Practicing [DADA2](https://benjjneb.github.io/dada2/index.html) pipeline to analyze amplcion data from Chinese Daqu samples.
* Using the [Silva database](https://www.arb-silva.de/) to annotate 16S sequences. 
* Examine bacterial genera relative abundance.
----
### Preparation of amplicon data (reads)
* Data were analyzed using Daqu samples from the CSBL.
----
### Link
* https://benjjneb.github.io/dada2/tutorial.html
